### ISSUE TITLE
Global regexOptions 

### ISSUE THREAD

**Issue by sm2017** [fayeed/flutter_parsed_text#38]
In version 2.2.0 you have a breaking change (Please follow semver)
> Removed regexOptions property from MatchText and now comes with a global regexOptions property for ParsedText Widget.

Why? I think it's totally wrong! Assume we have two `MatchText` , and want to set `caseSensitive: false` for first and  `caseSensitive: true` for second, how is it possible with a global regexOptions?

**Comment by fayeed** [comment]
The thing is the way it works previously I used to split the text into words and loop over those words with all the MatchText in the order they were passed, the problem with this approach was that regex would not work with a combination of words as they would get splited. The new approach works by generating a pattern from the MatchText which is then applied over the whole text, which resolves the above-mentioned issue. But the downside is that we cannot have a pattern with multiple regexOptions.

**Comment by sm2017** [comment]
>  I used to split the text into words and loop over those words

That means the first step is splitting the input text by space?

**Comment by fayeed** [comment]
yes

**Comment by sm2017** [comment]
So what happens if we have a regex that matches two word like `/[a-z]+ [0-9]+/`

**Comment by fayeed** [comment]
This wouldn't work with previous apporach

**Comment by sm2017** [comment]
How it works in new apporach?

**Comment by sm2017** [comment]
> by generating a pattern from the MatchText which
Can you show mw the code?

**Comment by fayeed** [comment]
https://github.com/fayeed/flutter_parsed_text/blob/7cf079be748832aa4c5c8a662078812e5ba59a45/lib/src/parsed_text.dart#L114

**Comment by sm2017** [comment]
I am too busy to fix the problem, But if anyone interested can read similar library like https://github.com/Khan/simple-markdown to get idea

**Comment by sm2017** [comment]
@fayeed by upgrading from `2.1.0` to ` 2.2.0` my code doesn't works

```
I/flutter ( 7570): ══╡ EXCEPTION CAUGHT BY WIDGETS LIBRARY ╞═══════════════════════════════════════════════════════════
I/flutter ( 7570): The following StateError was thrown building ParsedText(dirty):
I/flutter ( 7570): Bad state: No element
I/flutter ( 7570): 
I/flutter ( 7570): The relevant error-causing widget was:
I/flutter ( 7570):   ParsedText file:///C:/Projects/app/app/lib/widgets/formatted_text.dart:45:12
I/flutter ( 7570): 
I/flutter ( 7570): When the exception was thrown, this was the stack:
I/flutter ( 7570): #0      Iterable.firstWhere (dart:core/iterable.dart:516:5)
I/flutter ( 7570): #1      ParsedText.build.<anonymous closure> (package:flutter_parsed_text/src/parsed_text.dart:129:34)
I/flutter ( 7570): #2      _StringBase.splitMapJoin (dart:core-patch/string_patch.dart:824:27)
I/flutter ( 7570): #3      ParsedText.build (package:flutter_parsed_text/src/parsed_text.dart:118:15)
I/flutter ( 7570): #4      StatelessElement.build (package:flutter/src/widgets/framework.dart:4648:28)
I/flutter ( 7570): #5      ComponentElement.performRebuild (package:flutter/src/widgets/framework.dart:4574:15)
I/flutter ( 7570): #6      Element.rebuild (package:flutter/src/widgets/framework.dart:4267:5)
I/flutter ( 7570): #7      ComponentElement._firstBuild (package:flutter/src/widgets/framework.dart:4553:5)
I/flutter ( 7570): #8      ComponentElement.mount (package:flutter/src/widgets/framework.dart:4548:5)
I/flutter ( 7570): ...     Normal element mounting (19 frames)
I/flutter ( 7570): #27     Element.inflateWidget (package:flutter/src/widgets/framework.dart:3611:14)
I/flutter ( 7570): #28     MultiChildRenderObjectElement.inflateWidget (package:flutter/src/widgets/framework.dart:6221:36)
I/flutter ( 7570): #29     MultiChildRenderObjectElement.mount (package:flutter/src/widgets/framework.dart:6232:32)
I/flutter ( 7570): ...     Normal element mounting (22 frames)
I/flutter ( 7570): #51     Element.inflateWidget (package:flutter/src/widgets/framework.dart:3611:14)
I/flutter ( 7570): #52     MultiChildRenderObjectElement.inflateWidget (package:flutter/src/widgets/framework.dart:6221:36)
I/flutter ( 7570): #53     MultiChildRenderObjectElement.mount (package:flutter/src/widgets/framework.dart:6232:32)
I/flutter ( 7570): ...     Normal element mounting (142 frames)
I/flutter ( 7570): #195    Element.inflateWidget (package:flutter/src/widgets/framework.dart:3611:14)
I/flutter ( 7570): #196    Element.updateChild (package:flutter/src/widgets/framework.dart:3363:18)
I/flutter ( 7570): #197    SliverMultiBoxAdaptorElement.updateChild (package:flutter/src/widgets/sliver.dart:1241:37)
I/flutter ( 7570): #198    SliverMultiBoxAdaptorElement.createChild.<anonymous closure> (package:flutter/src/widgets/sliver.dart:1226:20)
I/flutter ( 7570): #199    BuildOwner.buildScope (package:flutter/src/widgets/framework.dart:2535:19)
I/flutter ( 7570): #200    SliverMultiBoxAdaptorElement.createChild (package:flutter/src/widgets/sliver.dart:1219:12)
I/flutter ( 7570): #201    RenderSliverMultiBoxAdaptor._createOrObtainChild.<anonymous closure> (package:flutter/src/rendering/sliver_multi_box_adaptor.dart:349:23)
I/flutter ( 7570): #202    RenderObject.invokeLayoutCallback.<anonymous closure> (package:flutter/src/rendering/object.dart:1889:59)
I/flutter ( 7570): #203    PipelineOwner._enableMutationsToDirtySubtrees (package:flutter/src/rendering/object.dart:912:15)
I/flutter ( 7570): #204    RenderObject.invokeLayoutCallback (package:flutter/src/rendering/object.dart:1889:14)
I/flutter ( 7570): #205    RenderSliverMultiBoxAdaptor._createOrObtainChild (package:flutter/src/rendering/sliver_multi_box_adaptor.dart:338:5)
I/flutter ( 7570): #206    RenderSliverMultiBoxAdaptor.addInitialChild (package:flutter/src/rendering/sliver_multi_box_adaptor.dart:422:5)
I/flutter ( 7570): #207    RenderSliverList.performLayout (package:flutter/src/rendering/sliver_list.dart:79:12)
I/flutter ( 7570): #208    RenderObject.layout (package:flutter/src/rendering/object.dart:1779:7)
I/flutter ( 7570): #209    RenderSliverEdgeInsetsPadding.performLayout (package:flutter/src/rendering/sliver_padding.dart:137:12)
I/flutter ( 7570): #210    RenderSliverPadding.performLayout (package:flutter/src/rendering/sliver_padding.dart:371:11)
I/flutter ( 7570): #211    RenderObject.layout (package:flutter/src/rendering/object.dart:1779:7)
I/flutter ( 7570): #212    RenderSliverEdgeInsetsPadding.performLayout (package:flutter/src/rendering/sliver_padding.dart:137:12)
I/flutter ( 7570): #213    RenderSliverPadding.performLayout (package:flutter/src/rendering/sliver_padding.dart:371:11)
I/flutter ( 7570): #214    RenderObject.layout (package:flutter/src/rendering/object.dart:1779:7)
I/flutter ( 7570): #215    RenderViewportBase.layoutChildSequence (package:flutter/src/rendering/viewport.dart:510:13)
I/flutter ( 7570): #216    RenderViewport._attemptLayout (package:flutter/src/rendering/viewport.dart:1574:12)
I/flutter ( 7570): #217    RenderViewport.performLayout (package:flutter/src/rendering/viewport.dart:1483:20)
I/flutter ( 7570): #218    RenderObject.layout (package:flutter/src/rendering/object.dart:1779:7)
I/flutter ( 7570): #219    RenderProxyBoxMixin.performLayout (package:flutter/src/rendering/proxy_box.dart:116:14)
I/flutter ( 7570): #220    RenderObject.layout (package:flutter/src/rendering/object.dart:1779:7)
I/flutter ( 7570): #221    RenderProxyBoxMixin.performLayout (package:flutter/src/rendering/proxy_box.dart:116:14)
I/flutter ( 7570): #222    RenderObject.layout (package:flutter/src/rendering/object.dart:1779:7)
I/flutter ( 7570): #223    RenderProxyBoxMixin.performLayout (package:flutter/src/rendering/proxy_box.dart:116:14)
I/flutter ( 7570): #224    RenderObject.layout (package:flutter/src/rendering/object.dart:1779:7)
I/flutter ( 7570): #225    RenderProxyBoxMixin.performLayout (package:flutter/src/rendering/proxy_box.dart:116:14)
I/flutter ( 7570): #226    RenderObject.layout (package:flutter/src/rendering/object.dart:1779:7)
I/flutter ( 7570): #227    RenderProxyBoxMixin.performLayout (package:flutter/src/rendering/proxy_box.dart:116:14)
I/flutter ( 7570): #228    RenderObject.layout (package:flutter/src/rendering/object.dart:1779:7)
I/flutter ( 7570): #229    RenderProxyBoxMixin.performLayout (package:flutter/src/rendering/proxy_box.dart:116:14)
I/flutter ( 7570): #230    RenderObject.layout (package:flutter/src/rendering/object.dart:1779:7)
I/flutter ( 7570): #231    RenderProxyBoxMixin.performLayout (package:flutter/src/rendering/proxy_box.dart:116:14)
I/flutter ( 7570): #232    RenderObject.layout (package:flutter/src/rendering/object.dart:1779:7)
I/flutter ( 7570): #233    RenderProxyBoxMixin.performLayout (package:flutter/src/rendering/proxy_box.dart:116:14)
I/flutter ( 7570): #234    RenderCustomPaint.performLayout (package:flutter/src/rendering/custom_paint.dart:545:11)
I/flutter ( 7570): #235    RenderObject.layout (package:flutter/src/rendering/object.dart:1779:7)
I/flutter ( 7570): #236    RenderProxyBoxMixin.performLayout (package:flutter/src/rendering/proxy_box.dart:116:14)
I/flutter ( 7570): #237    RenderObject.layout (package:flutter/src/rendering/object.dart:1779:7)
I/flutter ( 7570): #238    RenderProxyBoxMixin.performLayout (package:flutter/src/rendering/proxy_box.dart:116:14)
I/flutter ( 7570): #239    RenderObject.layout (package:flutter/src/rendering/object.dart:1779:7)
I/flutter ( 7570): #240    ChildLayoutHelper.layoutChild (package:flutter/src/rendering/layout_helper.dart:54:11)
I/flutter ( 7570): #241    RenderFlex._computeSizes (package:flutter/src/rendering/flex.dart:896:45)
I/flutter ( 7570): #242    RenderFlex.performLayout (package:flutter/src/rendering/flex.dart:931:32)
I/flutter ( 7570): #243    RenderObject.layout (package:flutter/src/rendering/object.dart:1779:7)
I/flutter ( 7570): #244    RenderProxyBoxMixin.performLayout (package:flutter/src/rendering/proxy_box.dart:116:14)
I/flutter ( 7570): #245    RenderObject.layout (package:flutter/src/rendering/object.dart:1779:7)
I/flutter ( 7570): #246    MultiChildLayoutDelegate.layoutChild (package:flutter/src/rendering/custom_layout.dart:171:12)
I/flutter ( 7570): #247    _ScaffoldLayout.performLayout (package:flutter/src/material/scaffold.dart:937:7)
I/flutter ( 7570): #248    MultiChildLayoutDelegate._callPerformLayout (package:flutter/src/rendering/custom_layout.dart:240:7)
I/flutter ( 7570): #249    RenderCustomMultiChildLayoutBox.performLayout (package:flutter/src/rendering/custom_layout.dart:404:14)
I/flutter ( 7570): #250    RenderObject._layoutWithoutResize (package:flutter/src/rendering/object.dart:1636:7)
I/flutter ( 7570): #251    PipelineOwner.flushLayout (package:flutter/src/rendering/object.dart:881:18)
I/flutter ( 7570): #252    RendererBinding.drawFrame (package:flutter/src/rendering/binding.dart:462:19)
I/flutter ( 7570): #253    WidgetsBinding.drawFrame (package:flutter/src/widgets/binding.dart:876:13)
I/flutter ( 7570): #254    RendererBinding._handlePersistentFrameCallback (package:flutter/src/rendering/binding.dart:328:5)
I/flutter ( 7570): #255    SchedulerBinding._invokeFrameCallback (package:flutter/src/scheduler/binding.dart:1144:15)
I/flutter ( 7570): #256    SchedulerBinding.handleDrawFrame (package:flutter/src/scheduler/binding.dart:1082:9)
I/flutter ( 7570): #257    SchedulerBinding._handleDrawFrame (package:flutter/src/scheduler/binding.dart:998:5)
I/flutter ( 7570): #261    _invoke (dart:ui/hooks.dart:163:10)
I/flutter ( 7570): #262    PlatformDispatcher._drawFrame (dart:ui/platform_dispatcher.dart:259:5)
I/flutter ( 7570): #263    _drawFrame (dart:ui/hooks.dart:126:31)
I/flutter ( 7570): (elided 3 frames from dart:async)
I/flutter ( 7570): 
I/flutter ( 7570): ════════════════════════════════════════════════════════════════════════════════════════════════════

======== Exception caught by widgets library =======================================================
The following StateError was thrown building ParsedText(dirty):
Bad state: No element

The relevant error-causing widget was: 
  ParsedText file:///C:/Projects/app/app/lib/widgets/formatted_text.dart:45:12
When the exception was thrown, this was the stack: 
#0      Iterable.firstWhere (dart:core/iterable.dart:516:5)
#1      ParsedText.build.<anonymous closure> (package:flutter_parsed_text/src/parsed_text.dart:129:34)
#2      _StringBase.splitMapJoin (dart:core-patch/string_patch.dart:824:27)
#3      ParsedText.build (package:flutter_parsed_text/src/parsed_text.dart:118:15)
#4      StatelessElement.build (package:flutter/src/widgets/framework.dart:4648:28)
...
====================================================================================================

```

**Comment by sm2017** [comment]
@fayeed to preproduce 


```
final parse = <MatchText>[
    MatchText(
      pattern: r"@([a-z][a-z0-9_]{4,31})",
      regexOptions: RegexOptions(caseSensitive: false),
      renderWidget: ({pattern, text}) => Text(
        text,
        textDirection: TextDirection.ltr,
        style: TextStyle(
          decoration: TextDecoration.underline,
        ),
      ),
      onTap: (String username) {
        resolveUsername(
          username.substring(1),
        );
      },
    ),
  ];
final defaultTextStyle = DefaultTextStyle.of(context);

return ParsedText(
      text: text,
      style: defaultTextStyle ,
      alignment:  TextAlign.start,
      textDirection: Directionality.of(context),
      overflow: TextOverflow.clip,
      maxLines: defaultTextStyle.maxLines,
      parse: parse,
    );

```

**Comment by fayeed** [comment]
You would see this error only if you have an empty parse list, ideally you shouldn't see the error if the list is empty so I will have to fix this.

**Comment by sm2017** [comment]
@fayeed As you can see the `parse` is not empty!!

**Comment by fayeed** [comment]
@sm2017 can you post a complete example here, from the example I saw that the match text list is empty and I just tested this & it only happens for me when the list is empty.

**Comment by sm2017** [comment]
```dart
import 'package:flutter/material.dart';
import 'package:flutter_parsed_text/flutter_parsed_text.dart';

class FormattedText extends StatelessWidget {
  final String text;
  final TextStyle style;
  final TextAlign textAlign;
  final TextDirection textDirection;
  final TextOverflow overflow;
  final int maxLines;

  final parse = <MatchText>[
    MatchText(
      pattern: r"@([a-z][a-z0-9_]{4,31})",
      renderWidget: ({pattern, text}) => Text(
        text,
        textDirection: TextDirection.ltr,
        style: TextStyle(
          decoration: TextDecoration.underline,
        ),
      ),
      onTap: (String username) {
        print(username.substring(1));
      },
    ),
  ];

  FormattedText(
    this.text, {
    Key key,
    this.style,
    this.textAlign,
    this.textDirection,
    this.overflow,
    this.maxLines,
  })  : assert(text != null),
        super(key: key);

  @override
  Widget build(BuildContext context) {
    final defaultTextStyle = DefaultTextStyle.of(context);

    return ParsedText(
      text: text,
      style: style ?? defaultTextStyle.style,
      alignment: textAlign ?? defaultTextStyle.textAlign ?? TextAlign.start,
      textDirection: textDirection ?? Directionality.of(context),
      overflow: overflow ?? defaultTextStyle.overflow ?? TextOverflow.clip,
      maxLines: maxLines ?? defaultTextStyle.maxLines,
      parse: parse,
      regexOptions: RegexOptions(caseSensitive: false),
    );
  }
}

```


and use `FormattedText("hello @FAYEED")`

**Comment by fayeed** [comment]
Just tested this cant reproduce the error I am on Flutter 2.2.0

**Comment by sm2017** [comment]
I'm in flutter 2.2.0 too and it occurred!!

**Comment by sm2017** [comment]
if I use `FormattedText("hello @fayeed")` instead of `FormattedText("hello @FAYEED")` its okay , Do you tested with uppercase?

**Comment by fayeed** [comment]
Ok, I see the problem the regex options are not being passed to the mapping function that's why you see the error, I will raise a fix in a couple of hours.

**Comment by fayeed** [comment]
New version available v2.2.1
